### PR TITLE
Un-Haunted Chemistry

### PR DESCRIPTION
**Pull request by CliffracerX**
## TEST MERGE THIS FIRST, PLEASE!
I've had this TM'd on my private server and gotten a good few of my chemistry main friends to try and break things, *thus far* all seems to be working but I don't trust it to be ready for production without the swarms of the truly robust that tend to main baseline TG trying to abuse it to hell and back.  If it survives them for a couple days without new bugs emerging, *then* I think it could be okay to fullmerge, but until then I beg of ye, maintainers: treat this like a nuclear bomb in PR form.

## About The Pull Request

Spooky month is over, and so is the spookiness happening in FermiChem since the recent revamps - issues began with #78637 and a cascade of errors arose in the subsequent PRs trying to resolve them (e.g, #78884, #78947, #79084) as the underlying logic was no longer playing nice.  This PR is my attempt to revert what broke whilst retaining the general fixes and improvements that were hidden in amongst the attempts to get chemistry playing nice again.

For specifics: #78637 has been partially reverted (restored VOLUME_MINIMUM and VOLUME_ROUNDING, reverted to using round over FLOOR in most cases, restored sanity check in equilibrium.dm that was commented out), largely full-reverted #78884, #78947, and #79084, and then put in some work to fix the bugs that had been initially introduced in #78637.

From my understanding, the issues that it introduced/exposed are as follows:

1. ``add_reagent`` had some latent issues from where it returned ``TRUE`` instead of ``amount``, which #79084 actually fixed.
    * I've retained those fixes.
2. A fairly significant quantity of ``round``s were replaced with ``FLOOR``s in holder.dm.   IIRC, floor is default behaviour for BYOND's built-in round, and whatever our FLOOR macro is doing isn't behaving as expected.
3. A fair number of checks throughout the code were either looking for <= 0 (unlikely with floating point leftovers) or strictly < CHEMICAL_QUANTIZATION_LEVEL, which would leave small amounts at or above QUANTIZATION_LEVEL leftover or extra.
4. Many of these checks were also only looking for *tiny leftover chem amounts* rather than small fractions too much, leading to some of the plumbing clogs & weird "ghost chems" when clearing beakers or getting tiny amounts extra when printing with the ChemMaster.  The biggest offender is in ``trans_id_to``, which I restored the final round-to-VOLUME_ROUNDING check for.
5. Reaction chambers were trying to pull in floating-point magic numbers' worth of chems & could very easily clog themselves with just a bit too much (e.g, 3.0002 of a chem it's supposed to have flat-3 of) or tiny amounts just below the intended (0.9998 Multiver in an inacusiate reaction chamber) - these were exposed by the increased granularity but have been around for a while.
    * I've added some sanity checking to ensure they don't do this anymore.

Proof of testing, via a chem factory producing Multiver, Antihol, Inacusiate, Oculine, Salicylic Acid (misnamed bottles because I'm tired), Libital, Oxandrolone, Airui, Atropine, Salbutamol, Blood, and Synthflesh.  All bottles are 50u, Synthflesh patches are 40u.  The _only_ strangeness on display is with the blood & synthflesh factories and I suspect that to be a skill issue on my part as the numbers are too big to be from floating-point weirdness.

https://github.com/tgstation/tgstation/assets/2958111/b876019a-9fab-4f96-b03c-2a8c9325fa59

image

In some setups, QUANTIZATION_LEVEL extras can sometimes still sneak in:

image

But reaction clamps always make sure the excess goes away before it can cause issues.

image

## Why It's Good For The Game

Chemistry has been in varying levels of weird for around a month now and even I, a comparatively novice chemist, have been getting bothered by the bugs - and that's to speak nothing of my friends who are chem-factory gods and *suffering* with the bugs.  This is my best attempt to fix the issues without breaking the good stuff that came as part of the reworks - and to throw in my own part in making chemistry a little more robust against mathematical strangeness.

## Changelog

:cl: CliffracerX
fix: We finally expunged the ghosts from the chemistry lab & pharmacy, please contact your local Chemplain if any more emerge
/:cl:

### CODE REVIEW

File: code/datums/components/plumbing/_plumbing.dm [hunk 1]
```diff
@@ -93,30 +93,23 @@
...
 		return
 	net = ducts[num2text(dir)]
-
-	//find all valid suppliers in the duct
-	var/list/valid_suppliers = list()
 	for(var/datum/component/plumbing/supplier as anything in net.suppliers)
 		if(supplier.can_give(amount, reagent, net))
 			valid_suppliers += supplier
+	// Need to ask for each in turn very carefully, making sure we get the total volume. This is to avoid a division that would always round down and become 0
+	var/targetVolume = reagents.total_volume + amount
```
Comment (jlsnow301) [inline code/datums/components/plumbing/_plumbing.dm:109]: Sorry to be that guy but I'd rather this left in snake case. There should just not be camel case in dm

File: code/datums/components/plumbing/reaction_chamber.dm [hunk 2]
```diff
@@ -10,28 +10,27 @@
 /datum/component/plumbing/reaction_chamber/can_give(amount, reagent, datum/ductnet/net)
 	. = ..()
 	var/obj/machinery/plumbing/reaction_chamber/reaction_chamber = parent
-	if(!. || !reaction_chamber.emptying || reagents.is_reacting)
+	if(!. || !reaction_chamber.emptying || reagents.is_reacting == TRUE)
```
Comment (jlsnow301) [inline code/datums/components/plumbing/reaction_chamber.dm:13]: why do we need the explicit boolean here?

File: code/modules/plumbing/plumbers/teleporter.dm [hunk 3]
```diff
@@ -45,7 +45,7 @@
 ///Transfer reagents and display a flashing icon
 /obj/machinery/plumbing/sender/proc/teleport_chemicals(obj/machinery/plumbing/receiver/R, amount)
 	flick(initial(icon_state) + "_flash", src)
-	reagents.trans_to(R, amount)
+	reagents.trans_to(R, amount, round_robin = TRUE)
```
Comment (jlsnow301) [inline code/modules/plumbing/plumbers/teleporter.dm:48]: Can you name this R variable better while you're here please

File: code/datums/components/plumbing/_plumbing.dm [hunk 4]
```diff
@@ -93,30 +93,23 @@
...
 	var/suppliersLeft = valid_suppliers.len
-	if(!suppliersLeft)
-		return
-
-	//take an equal amount from each supplier
-	var/currentRequest
-	var/target_volume = reagents.total_volume + amount
 	for(var/datum/component/plumbing/give as anything in valid_suppliers)
-		currentRequest = (target_volume - reagents.total_volume) / suppliersLeft
+		var/currentRequest = (targetVolume - reagents.total_volume) / suppliersLeft
```
Comment (jlsnow301) [inline code/datums/components/plumbing/_plumbing.dm:112]: Same as above- in fact you're encouraged to rid the entire file of camel case if you want, but not a requirement

**Review by jlsnow301 — COMMENTED** [review]
Just some style nits

File: code/datums/components/plumbing/_plumbing.dm (hunk 4 above)
Comment (CliffracerX) [reply code/datums/components/plumbing/_plumbing.dm]: Should have these resolved now.

**Comment by MrMelbert** [comment]
> A fairly significant quantity of rounds were replaced with FLOORs in holder.dm. IIRC, floor is default behaviour for BYOND's built-in round, and whatever our FLOOR macro is doing isn't behaving as expected.

If this is the case this is a bigger problem than just chemistry, and we should probably have a unit test that floor is flooring correctly

**Comment by SyncIt21** [comment]
This PR is a over reaction and most of your issues are mis represented.

> <= 0 (unlikely with floating point leftovers)

This check is actually valid if you round the number with a quantization constant. For example look at this macro it rounds with a quantization constant
https://github.com/tgstation/tgstation/blob/0cfacbb71dce9f6c5845a2d4500c3addc9de3807/code/__DEFINES/atmospherics/atmos_mob_interaction.dm#L14
And then here we do the `<= 0` check
https://github.com/tgstation/tgstation/blob/0cfacbb71dce9f6c5845a2d4500c3addc9de3807/code/modules/atmospherics/gasmixtures/gas_mixture.dm#L81

In the exact similar way instead of doing round i am just using `FLOOR` without a macro over here
https://github.com/tgstation/tgstation/blob/0cfacbb71dce9f6c5845a2d4500c3addc9de3807/code/modules/reagents/chemistry/holder.dm#L334-L336

If there is problem with `FLOOR` then just replace it with round and that should fix most of your issues. First try this.

**The other BIG problems with this PR**

https://github.com/tgstation/tgstation/blob/51d9c870f82f9ef40030a7942be1bc88de18e3ca/code/modules/reagents/chemistry/holder.dm#L549
NOO remove this asap. The multiplier is applied incorrectly this will just reopen
- #79083

https://github.com/tgstation/tgstation/blob/851cd8956814cce80a9693338b4f72385e39f3f4/code/datums/components/plumbing/_plumbing.dm#L118
This is what broke pluming COMPLETLY!!. Do you understand this is the core of plumbing and is what causes excess amounts of reagents to flow in!! this will reopen
- #78945
- #78890

https://github.com/tgstation/tgstation/blob/12f7f2588c50c1d8f9c3fb6e0134b99e32fdc377/code/modules/reagents/chemistry/holder.dm#L583
This transfers reagents equaly and not proportionally and will reopen these issues
-  #78920
- #78912

Overall this PR is a ticking timebomb and should be closed ASAP. If the problem is with small amounts of reagents not getting removed or anything related to plumbing then create an detailed issue and i will look into it personally but this PR will cause a big headache for all of us

**Comment by CliffracerX** [comment]
Taking cited issues point-by-point:

### Cryo Chambers
#79083 is indeed back, but I'm actively working on fixing it, alongside the problematic transfer behaviour below.

### Rounding Error Plumbing
image
image

Couldn't repro #78945 nor #78890 whatsoever.

image
image

### Round Robin
This indeed is back as well, though I'm at this point more curious when expected-behaviour *broke* such that it warranted removal.  Will redo the removal on it.  Going to have to dig farther back in diffs to figure out what changed out of curiosity, but that's not important right now.

### FLOOR inaccuracies
I didn't explain what I was seeing clearly enough, judging by this; FLOOR seemed to be either acting akin to CEIL *or* possibly rounding down to zero as expected *but* then still leaving small amounts of floating point cruft that would break checks that assumed it rounded cleanly to zero, hence my switching many of the checks to <= QUANTIZATION_LEVEL.  Equally, when I mention the oddities in the blood/synthflesh output, it wasn't the result of any of the round ops at all - it was large decimals like 2.37, VV confirmed them to be accurate to that well past QUANTIZATION_LEVEL, and they had no impact on factory function unlike the other precision issues.

**Comment by SyncIt21** [comment]
> https://github.com/tgstation/tgstation/issues/79083 is indeed back, but I'm actively working on fixing it, alongside the problematic transfer behaviour below.

The fix is simple don't revert the changes i did in #79084. The problem was cryo was not multiplying the reagents inside the body after transferring. The order in which i applied multiplier was wrong is what i fixed and you once again broke it. Also the `trans_to` proc with multiplier is not only used by cryo but in a lot of other places as well and you broke it again by reverting my changes here
Screenshot (348)

> Round Robin
> This indeed is back as well, though I'm at this point more curious when expected-behaviour broke such that it warranted removal. Will redo the removal on it.

Nothing was broken with this. Why i removed it was explained in this PR
- #78884
Read the 1st point and you will understand.

> Couldn't repro https://github.com/tgstation/tgstation/issues/78945 nor https://github.com/tgstation/tgstation/issues/78890 whatsoever.

My bad i didn't read the latest commits. Yup no problem here

> FLOOR inaccuracies
> I didn't explain what I was seeing clearly enough, judging by this; FLOOR seemed to be either acting akin to CEIL or possibly rounding down to zero as expected but then still leaving small amounts of floating point cruft that would break checks that assumed it rounded cleanly to zero

Give reproduction steps. Show where it breaks. No one has reported any issues on it so far i don't understand what problem you are referring to. Yes FLOOR will act like CEIL and you will get weirdness i literarily explained the problem with a different programming language it can't be helped. The reason i did FLOOR & not round directly is because i did not want the volumes to exceed the `maximum_volume` var which can actually cause problems.

All in all what is the point of this PR exactly? It sounds like you just don't like seeing 4 point floating decimal values even though it does not cause any problems. You said this in your statement

> **Why It's Good For The Game**
> chemistry has been in varying levels of weird for around a month now and even I, a comparatively novice chemist, have been getting bothered by the bugs - and that's to speak nothing of my friends who are chem-factory gods and suffering with the bugs. This is my best attempt to fix the issues without breaking the good stuff that came as part of the reworks - and to throw in my own part in making chemistry a little more robust against mathematical strangeness.

Yet you haven't given any examples on
- what bugs are you exactly referring too. If you are talking about floating point Inaccuracies i have already explained how i account for those in the previous comment. Show examples of how reactions break or stop working. 
- what do you mean by robust mathematical strangeness?

All this is so vague.

Also the screenshots you showed when analysing values in the VV menu is only expected. It does not round to 2 decimal places when displaying values there it will show you all the raw decimal digits which is expected. So even though in the VV menu if total volume is something like 49.999 it won't cause any bugs because i have added an offset of 0.01 before testing and thus prevents any bugs.

Close this PR and first create issues in the bug tracker so maintainers can actually label them as bugs. Also most ofthe code which i optimized for speed has also been removed. There is no need for this.

**Comment by SyncIt21** [comment]
> but you will only end up putting back my changes by the time this PR is fixed

See. you are just proving my point. Bringing it all back

Screenshot (349)

I'll help you with this PR, You will see how all the code changes I made has nothing to do with the problems you are concerned with

**1. Review**
https://github.com/tgstation/tgstation/blob/438d69e17de0c3038c55ad6c5028571b8368e2b7/code/modules/reagents/chemistry/holder.dm#L1298-L1315

Remove this code ASAP. This is the most inefficient stuff you will ever see and considering how hot this proc is we don't need this here.
 
If the reagent should be deleted it will call `del_reagent()`. This proc will use a for loop to again locate the reagent(even though it was already located by `update_total()`) 
https://github.com/tgstation/tgstation/blob/438d69e17de0c3038c55ad6c5028571b8368e2b7/code/modules/reagents/chemistry/holder.dm#L376-L382

And finally this proc will (drum roll please....) will again call back `update_total()`
https://github.com/tgstation/tgstation/blob/438d69e17de0c3038c55ad6c5028571b8368e2b7/code/modules/reagents/chemistry/holder.dm#L393

Which will again call `del_reagent()` which will again use a for loop which will again call `update_total()` you get the point. Bring back the code i made in this [commit](https://github.com/tgstation/tgstation/pull/78884/commits/58478d14b4fedd7c558adf05671463d8fcd4e7aa)

**2. Review**
https://github.com/tgstation/tgstation/blob/438d69e17de0c3038c55ad6c5028571b8368e2b7/code/datums/components/plumbing/_plumbing.dm#L78
Undo this change and the one made to `holder_full()` proc. Remember i said this

> if total volume is something like 49.999 it won't cause any bugs because i have added an offset of 0.01 before testing Thus prevents said bugs.

That's what this proc currently does to prevent the rounding errors
https://github.com/tgstation/tgstation/blob/8f29ff6e97707fdd50225c2beb4be6853f2c0c66/code/modules/reagents/chemistry/holder.dm#L1377-L1378

**3. Review**
Screenshot (350)
Bring back all this code. We need an early return so we can do less work if we have no suppliers.

**4 Review**
Screenshot (351)
Bring back all of this code. The current code is clean & will call `process_request()` only once but you are calling it multiple times making the code look overall ugly.

**5 Review**
Screenshot (352)
No put this code back. 0.01 is a way more safer margin than `CHEMICAL_QUANTIZATION_LEVEL *10`(which equals 0.001) this and everywhere else you made this change.

And this is just the beginning. By reverting all my PR's you are bringing back less efficient code and removing all the optimizations i made in them. Rather than working so hard only to bring back the code i made it's best to just close this PR and make an bug report and let me handle it from there.

**Comment by SyncIt21** [comment]
See #79478

**Comment by jlsnow301** [comment]
Syncit has been pouring energy into this and I would hate to see it reverted just because it's not fully there yet

**Comment by rdragan** [comment]
> Syncit has been pouring energy into this and I would hate to see it reverted just because it's not fully there yet

Why was the initial refactor merged without being TM'd at all, despite being hilariously broken? Why was it even needed in the first place?

I don't want to discourage SyncIt from contributing, I'm not even gonna say he bit off more than he can chew. The real issue and the reason stuff like this happens and will keep happening is a complete lack of a proper development methodology and spiteful ignorance of having any real integration process at all.

**Comment by SyncIt21** [comment]
> Why was the initial refactor merged without being TM'd at all, despite being hilariously broken?

Hilariously broken is an over exaggeration on your part. The issue the author is representing only occurs when large plumbing setups run for extended amounts of time(which players usually don't have time to make & so the problem occurs very rarely) and even when it does occur it still has very minute consequences. The terms they used such as "clog & plug" never happen & i still haven't been able to reproduce it till date.

And even though the author thinks this PR solves the problems he has stated, high chances are he actually hasn't. He tested this PR using just 1 plumbing setup (which i have also tested in the above linked PR and it works) and it works fine but what about other plumbing setups? What about other reactions? The sample size is so small to conclude this PR solves everything and so problem would still occur months down the line because floating point math is very difficult to solve and so a few rounding operations (be it in my PR or this) would only temporarily mitigate the problem till a different set of circumstances makes the problem re-emerge

Also reverting PR's and calling it a fix is such a novice & cliche way of solving problems. It undoes not only the PR author's work but also the maintainers who reviewed the PR work as well causing much more problems than it solves. 

@jlsnow301 can i request you to please close this PR. This PR reverts so many code optimizations that it would be just time consuming to put them back. I have already salvaged the important fixes from this PR and have even mentioned the author in my PR changelog so they can get the GBP. If they are interested they can review my PR and suggest changes there but as for this PR it is unsalvagable.

**Comment by jlsnow301** [comment]
the maintainer team concurs.

> The real issue and the reason stuff like this happens and will keep happening is a complete lack of a proper development methodology and spiteful ignorance of having any real integration process at all.

This is drastic. In a codebase so massive and antiquated in parts, even the most benign code changes will smoke out a bug of what seemed to be previously unrelated. You just fix them, and move on. There's no need for this type of PR war this has become.

**Pull request closed without merging**